src/Raven.Server/ServerWide/Commands/ClusterTransactionCommand.cs

```diff
@@ -190,15 +200,73 @@ public List<string> ExecuteCompareExchangeCommands(ClusterOperationContext conte
 
             return null;
         }
-        
+
+        private void EnsureAtomicDocumentWrites(ClusterOperationContext context, long index)
+        {
+            if (SerializedDatabaseCommands.TryGet(nameof(DatabaseCommands), out BlittableJsonReaderArray commands))
+            {
+                ClusterCommands ??= new List<ClusterTransactionDataCommand>();
+                foreach (BlittableJsonReaderObject dbCmd in commands)
+                {
+                    var cmdType = dbCmd[nameof(ClusterTransactionDataCommand.Type)].ToString();
+                    var docId = dbCmd[nameof(ClusterTransactionDataCommand.Id)].ToString();
+                    var atomicGuardKey = "rvn-atomic-guard-" + docId;
+                    var document = (BlittableJsonReaderObject)dbCmd[nameof(ClusterTransactionDataCommand.Document)];
+                    switch (cmdType)
+                    {
+                        case nameof(CommandType.PUT):
+                            if (document.TryGetWithoutThrowingOnError(Constants.Documents.Metadata.Key,
+                                    out BlittableJsonReaderObject metadata) == false ||
+                                metadata.TryGetWithoutThrowingOnError(Constants.Documents.Metadata.ClusterTransactionIndex,
+                                    out long putIndex) == false)
+                            {
+                                putIndex = 0; // new cmp xchg entry, so no need to take an action 
+                            }
+
+                            ClusterCommands.Add(new ClusterTransactionDataCommand
+                            {
+                                Type = CommandType.CompareExchangePUT,
+                                Id = atomicGuardKey,
+                                Index = putIndex,
+                                Document = context.ReadObject(new DynamicJsonValue {["Id"] = docId}, "cmp-xchg-content")
+                            });
+                            if (metadata != null) // just to be on the safe side, should never happen
```

Comment:
why this should never happen ?

Reply:
What would happened if the `metadata` is `null`, should we create a new `metadata`?

Reply:
If this happens, that means that the user submitted a document without metadata.
The only way that can happen is if they crafted a special request. I'm including this check here to avoid NRE in that case, but this isn't something likely.